### PR TITLE
Convert Modal view to React components

### PR DESCRIPTION
**Pull request by gabrielcowley**
_This PR has a staging branch, which means a production build of this branch exists and can be launched. [**Click here**](https://www.landmarker.io/staging/react_modal) to preview this PR._

This PR converts the Modal view which is currently a Backbone View into React components and integrates it into the `reactbridge`.

This PR is to be merged after [PR #138](https://github.com/menpo/landmarker.io/pull/138).

### CODE REVIEW

**Comment by jabooth** [comment]
@gabrielcowley can you pull and merge `ts_react` to clean up the diff here?

**Comment by gabrielcowley** [comment]
@jabooth Will do

File: src/ts/app/view/components/PromptModal.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    event.preventDefault()
+    let v = value
+    if (v) {
+        v = v.toLowerCase()
+    }
+    submit(v)
+    close()
+}
+
+export interface PromptModalProps {
```
Comment (jabooth) [inline src/ts/app/view/components/PromptModal.tsx]: here we can write 
```
export interface PromptModalProps extends ModalProps
```
so we don't have to redeclare common fields

File: src/ts/app/view/components/PromptModal.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import * as React from 'react'
+import { Modal } from './Modal'
```
Comment (jabooth) [inline src/ts/app/view/components/PromptModal.tsx]:
```ts
import { Modal, PromptModalProps } from './Modal'
```

File: src/ts/app/view/components/ConfirmModal.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react'
+import { Modal } from './Modal'
+
+function acceptOrReject(acceptReject: (() => void) | undefined, close: () => void): void {
```
Comment (jabooth) [inline src/ts/app/view/components/ConfirmModal.tsx]: maybe more natural to make `acceptReject` optional (which means we have to put it last)
```ts
function acceptOrReject(close: () => void, acceptReject?: () => void)): void 
```

File: src/ts/app/view/components/ConfirmModal.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+export function ConfirmModal(props: ConfirmModalProps) {
+    return (
+        <Modal isOpen={props.isOpen} close={props.close} modifiers={['Small']} key={props.key}
+        closable={props.closable === undefined ? true : props.closable} title={props.title}>
+            <div className="ConfirmDialog">
+                <p>{props.message}</p>
+                <div className="ConfirmActions">
+                    <div className="ConfirmAction--Yes" onClick={() => acceptOrReject(props.accept, props.close)}>Yes</div>
```
Comment (jabooth) [inline src/ts/app/view/components/ConfirmModal.tsx]: if we do as above, flip args here

File: src/ts/app/view/reactbridge.ts [hunk 5]
```diff
@@ -153,7 +153,7 @@ export class ReactBridge {
 
     renderHelp() {
         const props: HelpProps = {
-            isVisable: this.app.isHelpOverlayOn,
+            isVisible: this.app.isHelpOverlayOn,
             onClick: () => this.app.toggleHelpOverlay()
         }
         const help = Help(props)
```
Comment (jabooth) [inline src/ts/app/view/reactbridge.ts]: it seems there should be more extensive change to `reactbridge` to integrate the new React Modal (doesn't seem to currently be used?)

**Review by jabooth — CHANGES_REQUESTED** [review]
Few specific things, and seems the connection in `reactbridge` is missing. 

I think `src/ts/app/model/modals.ts` can be simplified a lot, but this will become clearer once we see the integration. 

Remember, React creates the illusion of re-rendering the entire tree on every state change. The top level Modal manager class can itself be a React component that has state which tracks what Modals are currently open. It should wire up the callbacks so that changes registered from modals (i.e. clicking the cancel button) will update the state of the main modal manger in addition to the action that we wish to have performed.

Maybe this would be clearer to discuss in person on Skype @gabrielcowley?

**Comment by gabrielcowley** [comment]
All modals have been converted to React. The Intro modal still exists in Backbone form and the old version is still used for when Landmarker is opened. Replacing this with its React version would require a significant overhaul of the startup process. I'm happy for this to be merged as it is.

**Comment by jabooth** [comment]
@gabrielcowley OK, in the interest of unblocking let's get this merged, I'll circle back and tidy up the intro stuff later.

**Pull request merged**